design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
-
-The control plane will use the `Download` channel to push the configuration to the agent. See
-the [Configuration Download](#configuration-download) section for details.
+not an issue. The gRPC runtime will handle the connection establishment and management. If an error occurs or the stream or connection is dropped, the connection must be reestablished.
 
-The agent uses the `Upload` command to upload large payloads, such as the nginx configuration, to the control plane. For
-our use case, we only need the ability to send the nginx configuration from the control plane to the agent. We do not
-need the agent to upload its configuration to the control plane. Ideally, we would like to disable this functionality in
-the agent.
+#### Further Requirements and Implementation Details
 
-For the full commander service definition, see
-this [file](https://github.com/nginx/agent/blob/main/sdk/proto/command_svc.proto).
+- The control plane will need to run a gRPC server for the agent to connect to.
+- When an agent connects to the control plane, the payload _should_ contain the hostname (pod name) and nginx instanceID of the registering agent. This can be used to keep track of all agents/pods that are connected.
+  - We need to be able to link an agent connection with a subscription. These are two different gRPC calls. `Subscribe` is where we actually send an nginx config to an agent. We need to ensure that we are sending the correct nginx config to the correct agent. Ideally we use metadata from the agent connection (maybe hostname, maybe we need middleware to extract token/uuid from grpc connection headers) and store that in the context. The context is then used in the `Subscribe` call, where we can extract the agent identifier and send the proper nginx config for that identifier.
 
-#### Command Message
+  Process: agent `Connects` to NGF. We get its identifier and pod name, add the identifier(s) to a context cache, track that connection, and create a subscription channel for it. Agent then `Subscribes`. The context passed in allows us to use the identifier to grab the proper subscription channel and listen on it. This channel will receive a `ConfigApplyRequest` when we have a new nginx config to write.
 
-Command Messages have the following structure:
+- If a single nginx deployment is scaled, we should ensure that all instances for that deployment receive the same config (and aren't treated as "different").
+- Each graph that the control plane builds internally should be directly tied to an nginx deployment.
+- Whenever the control plane sees an nginx instance become Ready, we send its config to it (it doesn't matter if this is a new pod or a restarted pod).
+- If no nginx instances exist, control plane should not send any configuration.
+- The control plane should check if a connection exists first before sending the config (all replicas, even non-leaders, should be able to send the config because the agent may connect to a non-leader).
```

Comment:
I'm assuming the leader will be in charge of writing status. Is it possible for that status to become out of sync with the nginx configuration for a particular deployment?

For example, let's say we have two data plane pods and two control plane pods:
- dp1 connects to cp1
- dp2 connects to cp2
- cp2 is leader

Can cp1 fall behind due to network latencies/errors causing dp1 to be running version 1 of the config and cp2 to write the status corresponding to version 2 of the config?

Or, can cp2 fall behind retrying status and cp1 is now _x_ number of versions ahead?

Reply:
Yeah I'm not sure what the cleanest answer is here...

If the leader is the only one that can write status, then it seems like the replicas would need to share information with each other.

Reply:
There's another option of only allowing the leader pod to accept traffic, so all nginx agents would connect only to the leader. I think this could be done by having the NGF readiness probe only return true if it's the leader. Non-leader pods would be marked as Unready.

This would reduce complexity of sending config and writing status updates. Only the leader would be doing anything. Maybe this should be our approach instead, assuming no downsides.